common/flatpak-dir.c

```diff
@@ -4018,19 +4022,231 @@ repo_get_remote_collection_id (OstreeRepo *repo,
+      if (!access_token)
+        {
+          g_prefix_error (error, "Unable to get API server access token: ");
+          return NULL;
+        }
+
+      token_url = g_strconcat (api_server_url,
+                               "/token?ref=",
+                               ref_to_fetch,
+                               "&repo_url=",
+                               remote_url,
+                               NULL);
+
+      token_url_checksum = g_compute_checksum_for_string (G_CHECKSUM_SHA1,
+                                                          token_url,
+                                                          strlen (token_url));
+
+      cache_dir = flatpak_ensure_system_user_cache_dir_location (error);
+      if (!cache_dir)
+        {
+          g_prefix_error (error, "Unable to get system cache dir: ");
+          return NULL;
+        }
+
+      token_cache_filename = g_strconcat (g_file_get_path (cache_dir),
+                                          "/api-server-token-",
+                                          token_url_checksum,
+                                          NULL);
+      token_cache_file = g_file_new_for_path (token_cache_filename);
+      token_cache_file_info = g_file_query_info (token_cache_file,
+                                                 G_FILE_ATTRIBUTE_TIME_MODIFIED,
+                                                 G_FILE_QUERY_INFO_NONE,
+                                                 cancellable, NULL);
+      g_file_info_get_modification_time (token_cache_file_info, &token_cache_mtime);
+      g_get_current_time (&current_time);
+
+      /* Tokens expires in one hour, we won't reuse tokens that are close
+       * to expire, that's why we check for at most 3400 seconds old tokens */
+      if (g_file_query_exists (token_cache_file, cancellable) &&
+          current_time.tv_sec - token_cache_mtime.tv_sec < 3400)
+        {
+          if (!g_file_load_contents (token_cache_file,
+                                     cancellable,
+                                     &token_buffer,
+                                     NULL, NULL,
+                                     error))
+            {
+              g_warning ("Unable to read token cache file %s: %s",
+                         token_cache_filename,
+                         (*error)->message);
+              g_clear_pointer (&token_buffer, g_free);
+            }
+        }
+
+      if (token_buffer)
+        return token_buffer;
+
+      ensure_soup_session (dir);
+      request = soup_session_request_http (dir->soup_session, "GET",
+                                           token_url, error);
+      if (!request)
+        {
+          g_prefix_error (error, "Unable to create repo token Soup request: ");
+          return NULL;
+        }
+
+      http_message = soup_request_http_get_message (request);
+      bearer_header = g_strconcat ("Bearer: ", access_token, NULL);
```

Comment:
As I understand it, the [`Authorization` header doesn’t require an additional `:`](https://tools.ietf.org/html/rfc1945#section-11), so this should be `"Bearer "`.

This `access_token` is the same one which just came straight out of libsecret, so is presumably a fairly permanent secret which is shared between the device and the server. That seems like a bad idea: if anyone ever got a copy of the `access_token` for a particular machine, they could use it in perpetuity to access all the apps this user has access to. Instead, the code should send a fresh hash of the shared secret — something along the lines of `hmac(payload=current_date, key=access_token)`, but with a little more thought put into it than my quick sketch here.

Unless `access_token` is something set by some Hack-specific service on the machine, and rotated regularly? Basically I don’t think I know enough about the architecture here to comment for sure, but my ramblings above should outline what I think should be avoided.

Reply:
It's refreshed by the hack account service and has its own lifecycle etc which is maintained for us. In place of the hack token it could just as easily be any other OAuth thing from GOA or similar.

Reply:
That sounds appropriate then. Since it’s a (short life) shared secret, rather than a challenge and response, it can only be sent over HTTPS. So we should probably error out if the repository server returns a non-HTTPS URI for the token server.